src/main/java/com/google/devtools/build/lib/rules/cpp/CcCompilationHelper.java

```diff
@@ -950,13 +970,287 @@ private ImmutableSet<Artifact> getSourceArtifactsByType(
+    ImmutableMap<Artifact, Artifact> ddiFileMap = ddiFileMapBuilder.build();
+    var allModuleFilesBuilder =
+        NestedSetBuilder.<Artifact.DerivedArtifact>stableOrder().addAll(moduleFileMap.values());
+    var depModuleFiles = ccCompilationContext.getModuleFiles(usePic);
+    if (depModuleFiles != null) {
+      allModuleFilesBuilder.addTransitive(depModuleFiles);
+    }
+    var allModuleFiles = allModuleFilesBuilder.build();
+    for (CppSource source : moduleInterfaceSources.values()) {
+      Artifact sourceArtifact = source.getSource();
+
+      Label sourceLabel = source.getLabel();
+      CppCompileActionBuilder builder = initializeCompileAction(sourceArtifact);
+      builder.setPicMode(usePic);
+      builder.setActionName(CppActionNames.CPP20_MODULE_COMPILE);
+
+      builder
+          .addMandatoryInputs(additionalCompilationInputs)
+          .addAdditionalIncludeScanningRoots(additionalIncludeScanningRoots);
+
+      boolean bitcodeOutput =
+          featureConfiguration.isEnabled(CppRuleClasses.THIN_LTO)
+              && CppFileTypes.LTO_SOURCE.matches(sourceArtifact.getFilename());
+
+      String outputName = outputNameMap.get(sourceArtifact);
+      Artifact.DerivedArtifact moduleFile = moduleFileMap.get(sourceArtifact);
+      Preconditions.checkNotNull(moduleFile);
+      builder.setAdditionalOutputs(ImmutableList.of(moduleFile));
+      builder.setModuleFiles(allModuleFiles);
+      if (usePic) {
+        result.addPicCpp20ModuleFile(moduleFile);
+      } else {
+        result.addCpp20ModuleFile(moduleFile);
+      }
+
+      // all -fmodule-file=<module-name>=<path/to/bmi> flags are put in .modmap file
+      var modmapFile =
+          CppHelper.getCompileOutputArtifact(
+              actionConstructionContext,
+              label,
+              CppHelper.getArtifactNameForCategory(
+                  ccToolchain,
+                  ArtifactCategory.CPP_MODULES_MODMAP,
+                  getOutputNameBaseWith(outputName, usePic)),
+              configuration);
+      // all path/to/bmi are put in .modmap.input file,
+      // which is convenient to get all bmi in CppCompileAction
+      var modmapInputFile =
+          CppHelper.getCompileOutputArtifact(
+              actionConstructionContext,
+              label,
+              CppHelper.getArtifactNameForCategory(
+                  ccToolchain,
+                  ArtifactCategory.CPP_MODULES_MODMAP_INPUT,
+                  getOutputNameBaseWith(outputName, usePic)),
+              configuration);
+      var ddiFile = ddiFileMap.get(sourceArtifact);
+      Preconditions.checkNotNull(ddiFile);
+      createGenModmapAction(ddiFile, modulesInfoFile, modmapFile, modmapInputFile);
+      builder.setModmapFile(modmapFile);
+      builder.setModmapInputFile(modmapInputFile);
+
+      PathFragment ccRelativeName = sourceArtifact.getRootRelativePath();
+      // TODO(plf): Continue removing CLIF logic from C++. Follow up changes would include
```

Comment:
We can drop all the CLIF stuff since we know that our source is a module interface file.